### DISCUSSION TITLE
why are the Y and Z axis different in JSCAD vs. THREE?

### DISCUSSION THREAD

**Discussion by tsdexter** [jscad/OpenJSCAD.org#1292] [General]
@z3dev why are the axis' transposed vs THREEjs? 

For example, rendering the `primitives3D` example in each has them oriented differently with Z pointing upwards in JSCAD whereas Y points upwards in THREE: 

![image](https://github.com/jscad/OpenJSCAD.org/assets/4126987/ab88702c-48d8-4bc3-b119-839627d879d8)

**Comment by tsdexter** [comment]
@hrgdavor did you come across this when messing with THREE as well?

**Reply by hrgdavor** [reply]
it is simply the choice of default UP direction, none is correct/incorrect. If you want your scene in THREE to match jscad orientation just use
```js
camera.up.set(0, 0, 1)
```

... also in babylonjs the default triangle orientation is different than threejs CW/CCW choice is one more of those things where there is a chance some use one or the other and there is no definitive right choice as far as I know.

**Reply by tsdexter** [reply]
great thanks, I'll try that out tomorrow.

**Comment by tsdexter** [comment]
@hrgdavor @z3dev can you point me in the right direction for how the `transforms` property on the models works? For example: 

```js
console.log(rotate([degToRad(45), 0, 0], cuboid({size: [1,2,10], center: [2, 2, 2]})))
```
Logs: 
```js
{
  polygons: ...
  transforms:  [1, 0, 0, 0, 0, 0.7071067811865476, 0.7071067811865475, 0, 0, -0.7071067811865475, 0.7071067811865476, 0, 0, 0, 0, 1]
}
```

What do the 16 values represent? 

I was using `0, 5, 10` as the scale and `12, 13, 14` as the position determined via basic inspection of the values. That seemed to be working with basic cubes/translates/etc. Then I added some rotations and it's clearly not accurate. 

How can I accurately map the transforms array to data like position, scale, rotation for use with a three mesh or elsewhere?

**Reply by hrgdavor** [reply]
in jscadui there is compatibility layer for threejs, and in there is snippet how to apply transform from jscad
https://github.com/hrgdavor/jscadui/blob/2f1f8b73b235b16d80f9ff1bdd64bcd8090ae10e/packages/format-threejs/index.js#L88

jscadui supports multiple render engines so it first converts to common format and then to threejs or babylon or regl

common format
https://github.com/hrgdavor/jscadui/tree/main/packages/format-common

jsacd to common:
https://github.com/hrgdavor/jscadui/blob/main/packages/format-jscad/index.js

common to threejs
https://github.com/hrgdavor/jscadui/tree/main/packages/format-threejs

**Reply by tsdexter** [reply]
Nevermind. I see that it is a transformation matrix which I can pass to directly to `mesh.matrix` and turn off `mesh.matrixAutoUpdate` which seems to be working!

**Reply by tsdexter** [reply]
@hrgdavor didn't see your reply. Thanks - I'll look through the links provided to get a little more insight into how it's working!

**Reply by tsdexter** [reply]
@hrgdavor I can't use three controls with `mesh.matrixAutoUpdate=false` so I will have to look further at what you did to apply the JSCad transformations to the existing three matrix. Thanks!

**Reply by tsdexter** [reply]
@hrgdavor thanks! I got it working with: 

```js
meshRef.applyMatrix4({elements: model.transforms})
```
